### ISSUE TITLE
Desktop Integration

### ISSUE THREAD

**Issue by aschaeffer** [inexor-game/code#138]
(TODO)

**Comment by aschaeffer** [comment]
Added an description for the issue

**Comment by Croydon** [comment]
In the context of the Windows platform I don't really get the following two points:
> minimize the application (to tray or launcher) instead of switching to windowed mode
> launching applications (windows start button, unity launcher, gnome menus, ...)

Could you please describe what you mean? The "windows start button" sound like a normal application link? That's getting created by the Sauer installer. Do we use another installer?
And to the first point: When I'm alt+tab out of the game (or by using the windows key) it's working as expected, Sauer doesn't switch to windowed mode.

**Comment by aschaeffer** [comment]
Yes, that's a linux issue. You can leave the fullscreen mode and enter the windows mode, and then you're able to minimize. But you cannot minimize from fullscreen directly.

You're right that the links should be created by the installer. I'm not sure: do we have an installer for windows, already?

**Comment by Calinou** [comment]
A reminder: Windows start menu entries must go to their own folder. That is, `%PROGRAMDATA%\Microsoft\Windows\Start Menu\Programs\Inexor\Inexor.lnk` and not `%PROGRAMDATA%\Microsoft\Windows\Start Menu\Programs\Inexor.lnk`

**Comment by Croydon** [comment]
Btw, Sauerbraten is using on Windows this: http://nsis.sourceforge.net

**Comment by Croydon** [comment]
Similar discussion on Red Eclipse: https://github.com/red-eclipse/base/issues/175

**Comment by freezurbern** [comment]
> Btw, Sauerbraten is using on Windows this: http://nsis.sourceforge.net

quick heads up: Red Eclipse has a more up to date NSIS installer script [(Link)](https://github.com/red-eclipse/base/tree/master/src/install/win), if that's the route you want to take for Windows installations.

**Comment by Fohlen** [comment]
@freezurbern does this smoothly integrate with your CI/Travis (as on Windows?). We'd love to have some automatic packaging going on, which is why I'd suggest either the [Visual Studio Setup Extension](https://visualstudiogallery.msdn.microsoft.com/f1cc3f3e-c300-40a7-8797-c509fb8933b9) or, prefferably the [Chocolate package manager](https://chocolatey.org/)..

**Comment by freezurbern** [comment]
I have successfully built NSIS packages through my Jenkins instance using
the NSIS command line tools. It works great: on successful windows compile,
generate windows installer. I can show you the console scripts if you'd
like. Please note this is with Red Eclipse and the NSIS script i helped
with for them (linked earlier)

**Comment by aschaeffer** [comment]
Most of the desktop integration functionality will be provided by Inexor Flex.

### Icon

We have a [scalable version](https://github.com/inexor-game/artwork/blob/master/logo/logo_thick.svg) of the Inexor logo:

Usage:

* Window icon
* Application icon
* Start menu entry icon
* System tray icon

### Window icon

* https://wiki.libsdl.org/SDL_SetWindowIcon

### Start menu entries / Application icon

* Windows:
  * Link from the start menu folder to the BAT file
  * %PROGRAMDATA%\Microsoft\Windows\Start Menu\Programs\Inexor\Inexor.lnk
  * The .lnk file contains the icon (see start menu entries)
* Linux:
  * The icon location is: /usr/share/pixmaps
  * The .desktop file location is: /usr/share/applications or /usr/local/share/applications
  * The .desktop file references the icon and the shell script (inexor.sh)
  * https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html
  * https://help.ubuntu.com/community/UnityLaunchersAndDesktopFiles
  * https://wiki.archlinux.org/index.php/Desktop_entries

### The system tray functionality

* Shows that Inexor Flex is running
* Menu for starting Inexor Core instances (servers and clients)
* Minimize/restore the Inexor Core (Client) window
* Open the Inexor Flex user interface in the default web browser

Ubuntu Unity allows to integrate most of the system tray functionality in a start menu entry:

* https://help.ubuntu.com/community/UnityLaunchersAndDesktopFiles (Adding shortcuts to a launcher)

### Libraries

Though the functionality of the tray icons and the system notifications are almost the same different platforms have different APIs and locations. I would prefer a cross platform way to solve these problems. Therefore I did a little research which NodeJS libraries provides a cross platform support for desktop integration tasks and are appropriate for integration in Inexor Flex. 

* Desktop Notifications:
  * Cross platform: [node-notifier](https://github.com/mikaelbr/node-notifier)
  * Uses the notification center for macOS, notify-osd/libnotify-bin for Linux, toasters for windows 8/10 or taskbar balloons for earlier windows versions
* Tray Icon
  * Haven't found a cross platform library for NodeJS (only for [Java](https://github.com/dorkbox/SystemTray) and [Go](https://github.com/cratonica/trayhost) )
  * Windows: https://github.com/jsdnxx/tray-windows
  * OSX: https://github.com/brandonhorst/node-tray
  * Linux: no NodeJS library available

### Window control functionality

* Minimize / restore the window
* Fullscreen

Control the window of an Inexor Core instance using Inexor Flex:

<pre>
    ./flex minimize 31415
    ./flex maximize 31415
    ./flex fullscreen 31415
</pre>

**Comment by aschaeffer** [comment]
Node-JS Library for dbus: https://github.com/Shouqun/node-dbus
